apps/vscode/src/providers/editor/toggle.ts

```diff
@@ -12,16 +12,88 @@
  * AGPL (http://www.gnu.org/licenses/agpl-3.0.txt) for more details.
  *
  */
-
+import * as vscode from "vscode";
 import { commands, window, workspace, TextDocument, ViewColumn } from "vscode";
+import * as path from 'path';
+import * as quarto from "quarto-core";
+import fs from "node:fs";
+import yaml from "js-yaml";
 import { Command } from "../../core/command";
 import { isQuartoDoc, kQuartoLanguageId } from "../../core/doc";
 import { VisualEditorProvider } from "./editor";
 
 
+export async function determineMode(doc: string): Promise<string | undefined> {
+  let editorOpener = undefined;
+
+  // check if file itself has a mode
+  if (hasEditorMode(doc, "source")) {
+    editorOpener = "source";
+  }
+  else if (hasEditorMode(doc, "visual")) {
+    editorOpener = "visual";
+  }
+  // check if has a _quarto.yml or _quarto.yaml file with editor specified
+  else {
+    editorOpener = workspaceHasQuartoYaml();
+  }
+
+  return editorOpener;
+}
 
+export async function setEditorOpener() {
+  const config = vscode.workspace.getConfiguration('quarto').get<string>('defaultEditor');
+  const viewType = config === 'visual' ? VisualEditorProvider.viewType : 'textEditor';
+  vscode.workspace.getConfiguration('workbench').update('editor.defaultView', viewType, true);
+  await vscode.commands.executeCommand("workbench.action.setDefaultEditor",
+    vscode.Uri.file('filename.qmd'),
+    viewType
+  );
+}
+
+export function workspaceHasQuartoYaml() {
+  const workspaceFolders = vscode.workspace.workspaceFolders;
+
+  if (workspaceFolders && workspaceFolders.length > 0) {
+    const rootPath = workspaceFolders[0].uri.fsPath;  // Only look in the root directory of the first workspace folder
+
+    const quartoFilePathYml = path.join(rootPath, '_quarto.yml');
+    const quartoFilePathYaml = path.join(rootPath, '_quarto.yaml');
+
+    let fileContent: string | null = null;
+
+    if (fs.existsSync(quartoFilePathYml)) {
+      fileContent = fs.readFileSync(quartoFilePathYml, 'utf8');
+    } else if (fs.existsSync(quartoFilePathYaml)) {
+      fileContent = fs.readFileSync(quartoFilePathYaml, 'utf8');
+    }
 
-export function editInVisualModeCommand() : Command {
+    if (fileContent) {
+      const parsedYaml = yaml.load(fileContent) as any;
+      if (parsedYaml.editor === 'visual' || parsedYaml.editor === 'source') {
+        return parsedYaml.editor;
+      }
+    }
+  }
+
+  return undefined;
+}
+
+export function hasEditorMode(doc: string, mode: string) {
+  if (doc) {
+    const match = doc.match(quarto.kRegExYAML);
+    if (match) {
+      const yaml = match[0];
+      return (
+        !!yaml.match(new RegExp("^editor:\\s+" + mode + "\\s*$", "gm")) ||
+        !!yaml.match(new RegExp("^[ \\t]*" + mode + ":\\s*(default)?\\s*$", "gm"))
+      );
```

Comment:
Oh I just remembered there is another way to specify source vs visual in the yaml that applies when you specify extra options, so we need to handle that too

<img width="191" alt="Screenshot 2024-10-22 at 4 35 24 PM" src="https://github.com/user-attachments/assets/a620ae31-0f3f-4c08-94ac-e5eeac012878">

Reply:
We should really offer better facilities for editors to verify metadata. This works well enough for now, but it could come from a `_metadata.yml` file a couple of directories up, or a `_quarto.yml` configuration, or ...

Reply:
`quarto inspect` probably works, but I bet it's too slow to be usable without a large degree of annoyance.

Reply:
Having some sort of toolkit around the wrangling all the places/configurations the yaml metadata might be set would be super helpful! I've never used `quarto inspect` before. Trying it locally seems to take multiple seconds for a single, simple file, so I would +1 the "too slow to be usable" idea 😅